contracts/utils/EnumerableMap.sol

```diff
@@ -0,0 +1,182 @@
+pragma solidity ^0.6.0;
+
+library EnumerableMap {
+    // To implement this library for multiple types with as little code
+    // repetition as possible, we write it in terms of a generic Map type with
+    // bytes32 keys and values.
+    // The Map implementation uses private functions, and user-facing
+    // implementations (such as Uint256ToAddressMap) are just wrappers around
+    // the underlying Map.
+    // This means that we can only create new EnumerableMaps for types that fit
+    // in bytes32.
+
+    struct MapEntry {
+        bytes32 _key;
+        bytes32 _value;
+    }
+
+    struct Map {
+        // Storage of map keys and values
+        MapEntry[] _entries;
+
+        // Position of the entry defined by a key in the `entries` array, plus 1
+        // because index 0 means a key is not in the map.
+        mapping (bytes32 => uint256) _indexes;
+    }
+
+    /**
+     * @dev Add a key-value pair to a map. O(1).
+     *
+     * If the key was already in the map, its value is updated and returns
+     * false.
+     */
+    function add(Map storage map, bytes32 key, bytes32 value) private returns (bool) {
+        if (!contains(map, key)) {
+            map._entries.push(MapEntry({ _key: key, _value: value }));
+            // The entry is stored at length-1, but we add 1 to all indexes
+            // and use 0 as a sentinel value
+            map._indexes[key] = map._entries.length;
+            return true;
+        } else {
+            map._entries[map._indexes[key] - 1]._value = value;
```

Comment:
It's silly because this is a simple optimization the compiler could make :( But yes, we can improve it at the expense of making the code a bit more fragile.